PyPI: Changes to the Warehouse api

### Are you experiencing an issue with...

badge-maker NPM package

### 🐞 Description

[The Warehouse API](https://warehouse.pypa.io/api-reference/json.html) is changing and will remove the key `releases` from the response.

This change is breaking service tests for PyPI.

### 🔗 Link to the badge

_No response_

### 💡 Possible Solution

Refactor `services/pypi/pypi-base.js` and `services/pypi/pypi-helpers.js` to use information from `urls` not from releases

Thanks @ericof! To add some additional detail, it seems that all of our badges (and service tests) that are using the version-specific endpoint are already failing:

e.g.
https://img.shields.io/pypi/djversions/djangorestframework/3.7.3
![](https://img.shields.io/pypi/djversions/djangorestframework/3.7.3)
(screenshot for posterity)
![image](https://user-images.githubusercontent.com/13042488/181140038-2f644d1d-7801-4589-b5e2-9af75ef10d4b.png)


https://app.circleci.com/pipelines/github/badges/daily-tests/1843/workflows/bdf0d5ca-6ffc-419e-a760-a82f9de96b94/jobs/2745/tests

```bash
PypiPythonVersions [live] python versions (no versions specified) [ GET /pyshp/1.2.12.json ]
[ GET /pyshp/1.2.12.json ]
/home/circleci/project/shields/core/service-test-runner/cli.js

AssertionError: message mismatch: expected 'invalid response data' to equal 'missing'
    at Function._expectField (file:///home/circleci/project/shields/core/service-test-runner/icedfrisby-shields.js:85:51)
    at IcedFrisbyNock.<anonymous> (file:///home/circleci/project/shields/core/service-test-runner/icedfrisby-shields.js:70:26)
    at IcedFrisbyNock.<anonymous> (node_modules/icedfrisby/lib/icedfrisby.js:954:10)
    at invokeNextHook (node_modules/icedfrisby/lib/icedfrisby.js:1003:24)
    at /home/circleci/project/shields/node_modules/icedfrisby/lib/icedfrisby.js:1017:7
    at new Promise (<anonymous>)
    at IcedFrisbyNock._runHooks (node_modules/icedfrisby/lib/icedfrisby.js:976:12)
    at IcedFrisbyNock.run (node_modules/icedfrisby/lib/icedfrisby.js:1276:20)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Context.<anonymous> (node_modules/icedfrisby/lib/icedfrisby.js:1348:9)
```

This aligns with the documentation about the `releases` key being removed on the `release` API:

https://warehouse.pypa.io/api-reference/json.html#release

> Previously this response included the releases key, which had the URLs for all files for every release of this project on PyPI. Due to stability concerns, this had to be removed from the release specific page, which now ONLY serves data specific to that release.

In my mind this not only necessitates urgent action to fix these badges, but also means we need to get in front of the badges that are using the [project](https://warehouse.pypa.io/api-reference/json.html#project) endpoint that likely makes up a far larger portion of the pypi badges we're currently serving.

Something that's unclear to me in these docs @ericof are the multiple references to the "Simple API", do you know what those are referring to?

--- a/services/pypi/pypi-base.js
+++ b/services/pypi/pypi-base.js
@@ -9,16 +9,11 @@ const schema = Joi.object({
     license: Joi.string().allow('').allow(null),
     classifiers: Joi.array().items(Joi.string()).required(),
   }).required(),
-  releases: Joi.object()
-    .pattern(
-      Joi.string(),
-      Joi.array()
-        .items(
-          Joi.object({
-            packagetype: Joi.string().required(),
-          })
-        )
-        .required()
+  urls: Joi.array()
+    .items(
+      Joi.object({
+        packagetype: Joi.string().required(),
+      })
     )
     .required(),
 }).required()

--- a/services/pypi/pypi-helpers.js
+++ b/services/pypi/pypi-helpers.js
@@ -88,16 +88,12 @@ function getLicenses(packageData) {
 }
 
 function getPackageFormats(packageData) {
-  const {
-    info: { version },
-    releases,
-  } = packageData
-  const releasesForVersion = releases[version]
+  const { urls } = packageData
   return {
-    hasWheel: releasesForVersion.some(({ packagetype }) =>
+    hasWheel: urls.some(({ packagetype }) =>
       ['wheel', 'bdist_wheel'].includes(packagetype)
     ),
-    hasEgg: releasesForVersion.some(({ packagetype }) =>
+    hasEgg: urls.some(({ packagetype }) =>
       ['egg', 'bdist_egg'].includes(packagetype)
     ),
   }

--- a/services/pypi/pypi-helpers.spec.js
+++ b/services/pypi/pypi-helpers.spec.js
@@ -164,34 +164,17 @@ describe('PyPI helpers', function () {
 
   test(getPackageFormats, () => {
     given({
-      info: { version: '2.19.1' },
-      releases: {
-        '1.0.4': [{ packagetype: 'sdist' }],
-        '2.19.1': [{ packagetype: 'bdist_wheel' }, { packagetype: 'sdist' }],
-      },
+      urls: [{ packagetype: 'bdist_wheel' }, { packagetype: 'sdist' }],
     }).expect({ hasWheel: true, hasEgg: false })
     given({
-      info: { version: '1.0.4' },
-      releases: {
-        '1.0.4': [{ packagetype: 'sdist' }],
-        '2.19.1': [{ packagetype: 'bdist_wheel' }, { packagetype: 'sdist' }],
-      },
+      urls: [{ packagetype: 'sdist' }],
     }).expect({ hasWheel: false, hasEgg: false })
     given({
-      info: { version: '0.8.2' },
-      releases: {
-        0.8: [{ packagetype: 'sdist' }],
-        '0.8.1': [
-          { packagetype: 'bdist_egg' },
-          { packagetype: 'bdist_egg' },
-          { packagetype: 'sdist' },
-        ],
-        '0.8.2': [
-          { packagetype: 'bdist_egg' },
-          { packagetype: 'bdist_egg' },
-          { packagetype: 'sdist' },
-        ],
-      },
+      urls: [
+        { packagetype: 'bdist_egg' },
+        { packagetype: 'bdist_egg' },
+        { packagetype: 'sdist' },
+      ],
     }).expect({ hasWheel: false, hasEgg: true })
   })
 })

--- a/services/pypi/pypi-license.tester.js
+++ b/services/pypi/pypi-license.tester.js
@@ -24,7 +24,7 @@ t.create('license (from trove classifier)')
           license: '',
           classifiers: ['License :: OSI Approved :: MIT License'],
         },
-        releases: {},
+        urls: [],
       })
   )
   .expectBadge({
@@ -46,7 +46,7 @@ t.create('license (as acronym from trove classifier)')
             'License :: OSI Approved :: GNU General Public License (GPL)',
           ],
         },
-        releases: {},
+        urls: [],
       })
   )
   .expectBadge({

--- a/services/pypi/pypi-version.tester.js
+++ b/services/pypi/pypi-version.tester.js
@@ -43,7 +43,7 @@ t.create('no trove classifiers')
           license: 'foo',
           classifiers: [],
         },
-        releases: {},
+        urls: [],
       })
   )
   .expectBadge({

